src/db.c

```diff
@@ -50,14 +51,53 @@ void updateLFU(robj *val) {
     val->lru = (LFUGetTimeInMinutes()<<8) | counter;
 }
 
-/* Low level key lookup API, not actually called directly from commands
- * implementations that should instead rely on lookupKeyRead(),
- * lookupKeyWrite() and lookupKeyReadWithFlags(). */
+/* Lookup a key for read or write operations, or return NULL if the key is not
+ * found in the specified DB. This function implements the functionality of
+ * lookupKeyRead(), lookupKeyWrite() and their ...WithFlags() variants.
+ *
+ * Side-effects of calling this function:
+ *
+ * 1. A key gets expired if it reached it's TTL.
+ * 2. The key's last access time is updated.
+ * 3. The global keys hits/misses stats are updated (reported in INFO).
+ * 4. If keyspace notifications are enabled, a "keymiss" notification is fired.
+ *
+ * Flags change the behavior of this command:
+ *
+ *  LOOKUP_NONE (or zero): No special flags are passed.
+ *  LOOKUP_NOTOUCH: Don't alter the last access time of the key.
+ *  LOOKUP_NONOTIFY: Don't trigger keyspace event on key miss.
+ *  LOOKUP_NOSTATS: Don't increment key hits/misses couters.
+ *  LOOKUP_WRITE: Prepare the key for writing (delete expired keys even on
+ *                replicas, use separate keyspace stats and events (TODO)).
+ *
+ * Note: this function also returns NULL if the key is logically expired but
+ * still existing, in case this is a replica and the LOOKUP_WRITE is not set.
+ * Even if the key expiry is master-driven, we can correctly report a key is
+ * expired on replicas even if the master is lagging expiring our key via DELs
+ * in the replication link. */
 robj *lookupKey(redisDb *db, robj *key, int flags) {
     dictEntry *de = dictFind(db->dict,key->ptr);
+    robj *val = NULL;
     if (de) {
-        robj *val = dictGetVal(de);
+        val = dictGetVal(de);
+        int force_delete_expired = flags & LOOKUP_WRITE;
+        if (force_delete_expired) {
+            /* Forcing deletion of expired keys on a replica makes the replica
+             * inconsistent with the master. The reason it's allowed for write
+             * commands is to make writable replicas behave consistently. It
+             * shall not be used in readonly commands. Modules are accepted so
+             * that we don't break old modules. */
+            client *c = server.in_eval ? server.lua_client : server.current_client;
+            serverAssert(!c || !c->cmd || (c->cmd->flags & (CMD_WRITE|CMD_MODULE)));
```

Comment:
this assert assume that opening a key for write can’t be called from a redis command
but in RedisGraph when a SAVE command end we delete the temporary keys we created during the save process
we need to be able to workaround this assertion
this crashes redis in our tests
stack trace example:

```
------ STACK TRACE ------                                                                  
                                                                                           
Backtrace:                                                                                 
0   redis-server                        0x000000010cad474a lookupKey.cold.1 + 26           
1   redis-server                        0x000000010c9f1902 lookupKey + 402                 
2   redis-server                        0x000000010ca6b704 RM_OpenKey + 68                 
3   redisgraph.so                       0x000000010d3391f2 _DeleteGraphMetaKeys + 194      
4   redisgraph.so                       0x000000010d339392 _ClearKeySpaceMetaKeys + 88     
5   redisgraph.so                       0x000000010d339597 _PersistenceEventHandler + 202  
6   redis-server                        0x000000010ca74b93 moduleFireServerEvent + 195     
7   redis-server                        0x000000010ca03fd1 rdbSave + 641                   
8   redis-server                        0x000000010ca08eaa saveCommand + 218               
9   redis-server                        0x000000010c9cddd6 call + 278                      
10  redis-server                        0x000000010c9cef08 processCommand + 2904   
```

Reply:
@zuiderkwast we didn't want to break modules, and assumed that `c->cmd` will be a module command, but with notifications and events, that could be done from other random contexts.

IIRC this assertion was just there in order to help us find native redis commands that are not flagged correctly.
it could have been sufficient to test that we're not on a writable-replica, but we thought that coverage for such a test will be low, and preferred to check our assumption on masters too.

As far as i can tell, our options now are:
1. remove that assert completely (as was argued before, right?)
2. make it run only on writable replicas
2. find another way to exclude modules

please share your thoughts, and remind me what i forgot.

Reply:
Just to be sure I understand what's happening: Is the `RM_OpenKey` triggered by something other than a module command, e.g. a keyspace notification or event, which is fired after some real command has executed?

If that's the case, can we set `c->cmd` to NULL before the firing the event? If we do, then this code will not appear to be part of any command, which I think is better than appearing as being run as part of some read-only command.

Another question: Can this ever happen on a readonly replica? If yes, then perhaps we should set `force_delete_expired` to false here to keep the replica consistent with its primary, rather than only bypassing the assert.

If we do, then I guess we can drop the assert entirely.

Reply:
From what i know of RedisGraph, when this code happens on a replica, should always be in a fork child, but it could still be from within a command.
i.e. when a SYNC command is received and triggers an immediate fork, the fork child process will have `c->cmd` still set on the stack.

What i don't like about nullifying `c->cmd` in the various event dispatches in module.c is that it's very far from the assertion.
i.e. we'll have to backup, nullify, and restore `c->cmd` and add some comments explaining that it's done to avoid an assertion on the other side of town.

Reply:
Agree. This assert looks at things far away too...

I think we can remove the assertion and never set `force_delete_expired` on a writable replica:

```C
int force_delete_expired = flags & LOOKUP_WRITE && !(server.masterhost && server.repl_slave_ro);
```

Reply:
@zuiderkwast please make a PR.